Default generated Cypress Component Testing setup errors out on `component-index.html` path

### Current Behavior

The default generated preset/cypress config setup for an Angular Cypress Component Testing setup on latest Nx fails on a library project. It creates an incorrect file path to the `component-index.html` file for Cypress to consume in the `cypress.config.ts`.

### Expected Behavior

The default generated Cypress Component Testing setup should work with an Angular Library Project.

### GitHub Repo

https://github.com/zjkipping/nx-cypress-component-testing

### Steps to Reproduce

1. `npm install` in the attached github repo
2. `npx nx component-test bar`
3. See error logs

1. checkout `working-fix` branch on same repo
2. `npx nx component-test bar`
3. See success logs


### Nx Report

```shell
>  NX   Report complete - copy this into the issue template

   Node   : 18.15.0
   OS     : win32-x64
   npm    : 9.5.0
   
   nx                 : 16.7.1
   @nx/js             : 16.7.1
   @nx/jest           : 16.7.1
   @nx/linter         : 16.7.1
   @nx/workspace      : 16.7.1
   @nx/angular        : 16.7.1
   @nx/cypress        : 16.7.1
   @nx/devkit         : 16.7.1
   @nx/eslint-plugin  : 16.7.1
   @nrwl/tao          : 16.7.1
   @nx/web            : 16.7.1
   @nx/webpack        : 16.7.1
   typescript         : 5.1.6
```


### Failure Logs

```shell
ERROR in   Error: Child compilation failed:
  Module not found: Error: Can't resolve 'C:\Users\zjkip\Documents\code\web\cypr  ess-component-testing\libs\bar\Users\zjkip\Documents\code\web\cypress-componen  t-testing\libs\bar\cypress\support\component-index.html' in 'C:\Users\zjkip\Do  cuments\code\web\cypress-component-testing'
  ModuleNotFoundError: Module not found: Error: Can't resolve 'C:\Users\zjkip\Do  cuments\code\web\cypress-component-testing\libs\bar\Users\zjkip\Documents\code  \web\cypress-component-testing\libs\bar\cypress\support\component-index.html'   in 'C:\Users\zjkip\Documents\code\web\cypress-component-testing'
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\Compilation.js:2022:28      
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\NormalModuleFactory.js:817:13
      at eval (eval at create (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:10:  1)
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\NormalModuleFactory.js:275:22
      at eval (eval at create (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:9:1  )
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\NormalModuleFactory.js:448:22
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\NormalModuleFactory.js:122:11
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\NormalModuleFactory.js:689:25
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\NormalModuleFactory.js:893:8
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\NormalModuleFactory.js:1013:5
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\neo-async\async.js:6883:13
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\webpack\lib\NormalModuleFactory.js:996:45
      at finishWithoutResolve (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\enhanced-resolve\lib\Resolver.js:372:11)
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\Resolver.js:461:15 
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\Resolver.js:519:5  
      at eval (eval at create (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:16:  1)
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\Resolver.js:519:5  
      at eval (eval at create (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:27:  1)
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\DescriptionFilePlugin.js:89:43
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\Resolver.js:519:5  
      at eval (eval at create (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:16:  1)
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\Resolver.js:519:5  
      at eval (eval at create (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:16:  1)
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\Resolver.js:519:5  
      at eval (eval at create (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:16:  1)
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\Resolver.js:519:5  
      at eval (eval at create (C:\Users\zjkip\Documents\code\web\cypress-compone  nt-testing\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:27:  1)
      at C:\Users\zjkip\Documents\code\web\cypress-component-testing\node_module  s\enhanced-resolve\lib\DescriptionFilePlugin.js:89:43
  
  - Compilation.js:2022 
    [cypress-component-testing]/[webpack]/lib/Compilation.js:2022:28
  
  - NormalModuleFactory.js:817 
    [cypress-component-testing]/[webpack]/lib/NormalModuleFactory.js:817:13
  
  
  - NormalModuleFactory.js:275 
    [cypress-component-testing]/[webpack]/lib/NormalModuleFactory.js:275:22
  
  
  - NormalModuleFactory.js:448 
    [cypress-component-testing]/[webpack]/lib/NormalModuleFactory.js:448:22
  
  - NormalModuleFactory.js:122 
    [cypress-component-testing]/[webpack]/lib/NormalModuleFactory.js:122:11
  
  - NormalModuleFactory.js:689 
    [cypress-component-testing]/[webpack]/lib/NormalModuleFactory.js:689:25
  
  - NormalModuleFactory.js:893 
    [cypress-component-testing]/[webpack]/lib/NormalModuleFactory.js:893:8
  
  - NormalModuleFactory.js:1013 
    [cypress-component-testing]/[webpack]/lib/NormalModuleFactory.js:1013:5
  
  - async.js:6883 
    [cypress-component-testing]/[neo-async]/async.js:6883:13
  
  - NormalModuleFactory.js:996 
    [cypress-component-testing]/[webpack]/lib/NormalModuleFactory.js:996:45
  
  - Resolver.js:372 finishWithoutResolve
    [cypress-component-testing]/[enhanced-resolve]/lib/Resolver.js:372:11
  
  - Resolver.js:461 
    [cypress-component-testing]/[enhanced-resolve]/lib/Resolver.js:461:15
  
  - Resolver.js:519 
    [cypress-component-testing]/[enhanced-resolve]/lib/Resolver.js:519:5
  
  
  - Resolver.js:519 
    [cypress-component-testing]/[enhanced-resolve]/lib/Resolver.js:519:5
  
  
  - DescriptionFilePlugin.js:89 
    [cypress-component-testing]/[enhanced-resolve]/lib/DescriptionFilePlugin.js:    89:43
  
  - Resolver.js:519 
    [cypress-component-testing]/[enhanced-resolve]/lib/Resolver.js:519:5
  
  
  - Resolver.js:519 
    [cypress-component-testing]/[enhanced-resolve]/lib/Resolver.js:519:5
  
  
  - Resolver.js:519 
    [cypress-component-testing]/[enhanced-resolve]/lib/Resolver.js:519:5
  
  
  - Resolver.js:519 
    [cypress-component-testing]/[enhanced-resolve]/lib/Resolver.js:519:5
  
  
  - DescriptionFilePlugin.js:89 
    [cypress-component-testing]/[enhanced-resolve]/lib/DescriptionFilePlugin.js:    89:43
  
  - child-compiler.js:169 
    [cypress-component-testing]/[html-webpack-plugin]/lib/child-compiler.js:169:    18
  
  - Compiler.js:551 finalCallback
    [cypress-component-testing]/[webpack]/lib/Compiler.js:551:5
  
  - Compiler.js:577 
    [cypress-component-testing]/[webpack]/lib/Compiler.js:577:11
  
  - Compiler.js:1200 
    [cypress-component-testing]/[webpack]/lib/Compiler.js:1200:17
  
  
  - task_queues:95 processTicksAndRejections
    node:internal/process/task_queues:95:5
  

2023-08-17 16:24:46: webpack compiled in 2205 ms (c15edf4faa78232a)
====================================================================================================
  (Run Starting)
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        12.17.4                                                                        │
  │ Browser:        Electron 106 (headless)                                                        │
  │ Node Version:   v18.15.0 (C:\Program Files\nodejs\node.exe)                                    │
  │ Specs:          1 found (bar.component.cy.ts)                                                  │
  │ Searched:       src/**/*.cy.ts, src/**/*.cy.js                                                 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
────────────────────────────────────────────────────────────────────────────────────────────────────

  Running:  bar.component.cy.ts                                                             (1 of 1)

  1) custom component should have hello header
  0 passing (351ms)
  1 failing
  1) custom component should have hello header:
     Error: No element found that matches selector [data-cy-root]. Please add a root element with data-cy-root attribute to your "component-index.html" file so that Cypress can attach your component to the DOM.
      at getContainerEl (webpack://angular/./node_modules/cypress/angular/dist/index.js:60:10)
      at CypressTestComponentRenderer.removeAllRootElements (webpack://angular/./node_modules/cypress/angular/dist/index.js:167:8)
      at CypressTestComponentRenderer.insertRootElement (webpack://angular/./node_modules/cypress/angular/dist/index.js:162:13)
      at TestBedImpl.createComponent (webpack://angular/./node_modules/@angular/core/fesm2022/testing.mjs:26895:30)
      at setupFixture (webpack://angular/./node_modules/cypress/angular/dist/index.js:215:33)
      at Context.mount (webpack://angular/./node_modules/cypress/angular/dist/index.js:294:20)
      at wrapped (http://localhost:8080/__cypress/runner/cypress_runner.js:137575:43)
      at <unknown> (http://localhost:8080/__cypress/runner/cypress_runner.js:143461:15)
      at tryCatcher (http://localhost:8080/__cypress/runner/cypress_runner.js:1807:23)
```


### Operating System

- [ ] macOS
- [ ] Linux
- [X] Windows
- [ ] Other (Please specify)

### Additional Information

_No response_

--- a/packages/angular/plugins/component-testing.ts
+++ b/packages/angular/plugins/component-testing.ts
@@ -104,15 +104,17 @@ ${e.stack ? e.stack : e}`
     ...nxBaseCypressPreset(pathToConfig, { testingType: 'component' }),
     // NOTE: cannot use a glob pattern since it will break cypress generated tsconfig.
     specPattern: ['src/**/*.cy.ts', 'src/**/*.cy.js'],
-    // cypress defaults to a relative path from the workspaceRoot instead of projectRoot
-    // set as absolute path in case this changes internally to cypress, this path isn't OS dependent
-    indexHtmlFile: joinPathFragments(
-      ctContext.root,
-      ctProjectConfig.root,
-      'cypress',
-      'support',
-      'component-index.html'
-    ),
+    // Cy v12.17.0+ does not work with aboslute paths for index file
+    // but does with relative pathing, since relative path is the default location, we can omit it
+    indexHtmlFile: requiresAbsolutePath()
+      ? joinPathFragments(
+          ctContext.root,
+          ctProjectConfig.root,
+          'cypress',
+          'support',
+          'component-index.html'
+        )
+      : undefined,
     devServer: {
       // cypress uses string union type,
       // need to use const to prevent typing to string
@@ -376,9 +378,7 @@ function isOffsetNeeded(
   ctProjectConfig: ProjectConfiguration
 ) {
   try {
-    const { version = null } = require('cypress/package.json');
-
-    const supportsWorkspaceRoot = !!version && gte(version, '12.9.0');
+    const supportsWorkspaceRoot = isCyVersionGreaterThanOrEqual('12.9.0');
 
     // if using cypress <v12.9.0 then we require the offset
     if (!supportsWorkspaceRoot) {
@@ -412,4 +412,35 @@ function isOffsetNeeded(
     // safest to assume we do
     return true;
   }
+}
+
+/**
+ * check if the cypress version is able to understand absolute paths to the indexHtmlFile option
+ * this is required for nx to work with cypress <v12.17.0 since the relative pathing is causes issues
+ * with invalid pathing.
+ * v12.17.0+ works with relative pathing
+ *
+ * if there is an error thrown then we assume it is an older version of cypress and use the absolute path
+ * as that was supported for longer.
+ *
+ * */
+function requiresAbsolutePath() {
+  try {
+    return !isCyVersionGreaterThanOrEqual('12.17.0');
+  } catch (e) {
+    if (process.env.NX_VERBOSE_LOGGING === 'true') {
+      logger.error(e);
+    }
+    return true;
+  }
+}
+
+/**
+ * Checks if the install cypress version is greater than or equal to the provided version.
+ * Does not catch errors as any custom logic for error handling is required on consumer side.
+ * */
+function isCyVersionGreaterThanOrEqual(version: string) {
+  const { version: cyVersion = null } = require('cypress/package.json');
+
+  return !!cyVersion && gte(cyVersion, version);
 }

